Code/GraphMol/RascalMCES/RascalResult.cpp

```diff
@@ -0,0 +1,817 @@
+    d_mcesMol.reset(new ROMol(*other.d_mcesMol));
+  }
+  return *this;
+}
+
+void RascalResult::largestFragOnly() { largestFragsOnly(1); }
+
+void RascalResult::largestFragsOnly(unsigned int numFrags) {
+  std::unique_ptr<RDKit::ROMol> mol1_frags(makeMolFrags(1));
+  // getMolFrags() returns boost::shared_ptr.  Ho-hum.
+  auto frags = RDKit::MolOps::getMolFrags(*mol1_frags, false);
+  if (numFrags < 1 || frags.size() < numFrags) {
+    return;
+  }
+  std::sort(frags.begin(), frags.end(),
+            [](const boost::shared_ptr<ROMol> &f1,
+               const boost::shared_ptr<ROMol> &f2) -> bool {
+              return f1->getNumAtoms() > f2->getNumAtoms();
+            });
+  frags.erase(frags.begin() + numFrags, frags.end());
+  rebuildFromFrags(frags);
+}
+
+void RascalResult::trimSmallFrags(unsigned int minFragSize) {
+  std::unique_ptr<RDKit::ROMol> mol1_frags(makeMolFrags(1));
+  // getMolFrags() returns boost::shared_ptr.  Ho-hum.
+  auto frags = RDKit::MolOps::getMolFrags(*mol1_frags, false);
+  for (auto &frag : frags) {
+    if (frag->getNumAtoms() < minFragSize) {
+      frag.reset();
+    }
+  }
+  frags.erase(std::remove_if(
+                  frags.begin(), frags.end(),
+                  [](const boost::shared_ptr<ROMol> &f) -> bool { return !f; }),
+              frags.end());
+  rebuildFromFrags(frags);
+}
+
+double RascalResult::getSimilarity() const {
+  if (!d_mol1 || !d_mol2) {
+    return 0.0;
+  }
+  return johnsonSimilarity(d_bondMatches, d_atomMatches, *d_mol1, *d_mol2);
+}
+
+void RascalResult::rebuildFromFrags(
+    const std::vector<boost::shared_ptr<ROMol>> &frags) {
+  // Force the re-creation of the SMARTS and other properties next time
+  // they-re needed.
+  d_smarts = "";
+  d_maxFragSep = -1;
+  d_ringNonRingBondScore = -1;
+  d_maxDeltaAtomAtomDist = -1;
+  d_largestFragSize = -1;
+
+  std::set<int> fragAtoms, fragBonds;
+  for (const auto &f : frags) {
+    for (auto atom : f->atoms()) {
+      if (atom->hasProp("ORIG_INDEX")) {
+        fragAtoms.insert(atom->getProp<int>("ORIG_INDEX"));
+      }
+    }
+    for (auto bond : f->bonds()) {
+      if (bond->hasProp("ORIG_INDEX")) {
+        fragBonds.insert(bond->getProp<int>("ORIG_INDEX"));
+      }
+    }
+  }
+  std::vector<std::pair<int, int>> newAtomMatches;
+  for (const auto &am : d_atomMatches) {
+    if (fragAtoms.find(am.first) != fragAtoms.end()) {
+      newAtomMatches.push_back(am);
+    }
+  }
+  d_atomMatches = newAtomMatches;
+  std::vector<std::pair<int, int>> new_bond_matches;
+  for (const auto &bm : d_bondMatches) {
+    if (fragBonds.find(bm.first) != fragBonds.end()) {
+      new_bond_matches.push_back(bm);
+    }
+  }
+  d_bondMatches = new_bond_matches;
+  d_numFrags = frags.size();
+  d_largestFragSize = frags.empty() ? 0 : frags.front()->getNumAtoms();
+}
+
+std::string RascalResult::createSmartsString() const {
+  if (!d_mol1 || !d_mol2) {
+    return "";
+  }
+  std::unique_ptr<RDKit::RWMol> smartsMol(new RDKit::RWMol);
```

Comment:
Does this really need to be a pointer? I think it's only used in this function, so it could just be an RWMol?